src/compiler/checker.ts

```diff
@@ -16285,7 +16285,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         let type = unionTypes.get(id);
         if (!type) {
             type = createType(TypeFlags.Union) as UnionType;
-            type.objectFlags = objectFlags | getPropagatingFlagsOfTypes(types, /*excludeKinds*/ TypeFlags.Nullable);
+            type.objectFlags = (objectFlags | getPropagatingFlagsOfTypes(types, /*excludeKinds*/ TypeFlags.Nullable)) & ~(ObjectFlags.IsUnknownLikeUnionComputed | ObjectFlags.IsUnknownLikeUnion);

```

Comment:
This is a naive~ fix for the issue. I think that it would be best to remove those object flags and use a different mechanism to cache this information. This would prevent accidental propagation like this as this information should be unique to any given type.

I'm just not sure what's the preferred place to keep this sort of information. Let me know what is your preferred solution for this kind of caching and I will adjust the PR.

cc @ahejlsberg

Reply:
Thanks for tracking this down. The call to `getUnionTypeFromSortedList` that causes the issue comes from `filterType`. I'd suggest changing that call to

```ts
return getUnionTypeFromSortedList(filtered, (type as UnionType).objectFlags & (ObjectFlags.PrimitiveUnion | ObjectFlags.ContainsIntersections), /*aliasSymbol*/ undefined, /*aliasTypeArguments*/ undefined, newOrigin);
```

i.e. filter the flags there and leave `getUnionTypeFromSortedList` alone. It might make sense to also change the parameter name from `objectFlags` to `precomputedObjectFlags` just to make its use clear. BTW, strictly speaking it isn't correct to reuse the old `ObjectFlags.ContainsIntersections` in `filterType` since the filtering operation may have removed all intersections. However, that flag is just an optimization hint, so no harm comes from propagating it.

Reply:
Regarding where to put the flags, I think they're fine where they are. We really don't want to introduce another property for this.

Reply:
You mention `PrimitiveUnion` and `ContainsIntersections` in the suggested change and in the following comment. I'm a little bit lost on how this relates to the reused `IsUnknownLikeUnionComputed` and `IsUnknownLikeUnion` that I have identified to be the root cause here.

Reply:
Those two flags are the only ones we want propagated into the new union type. All other flags should be zero. Including the ones you identified as the root cause.

Reply:
Ah, I see - that makes sense. I pushed out the requested change.